### ISSUE TITLE
puller should only sync with peers of PO >= storage radius

### ISSUE THREAD

**Issue by istae** [ethersphere/bee#3700]
@ldeffenb writes

> Ok, let's see if I can explain my issue with the current puller implementation.   It uses the kademlia depth to select peers to sync with.  So, if my node's overlay starts with 0x000 and the kademlia depth is 4, it will sync with peers from 0x0000 through 0x0fffff (leading 4 bits match).   Now, if the storage depth is 8, then it will sync bins 8 and higher from those nodes.  But my node really only wants chunks that start with 0x00... which is 8 bits matching my overlay.   But if I ask for bin 8 from one of my peers that starts with 0x0f, it will send me only chunks starting with 0x0f.. because that's what it has in its bin 8.   But I don't want those chunks in my reserve, so they go straight to the cache and get gc'd later.
And I do that with ALL of my peers that happen to start with 0x01 through 0x0f which I'm syncing because of my kademlia depth of 4.   And I ONLY ask those peers for bins greater than or equal to my storage radius, which is 8, which guarantees that 7/8 of the chunks that the puller pulls go into the cache and get gc'd.   And I won't get ANY 0x00 chunks from the 0x01 thorugh 0x0f peers because they have those chunks in lower numbered bins than my storage radius.
Did I explain that well enough?   And does it make sense?   It certainly fits my observations when the kademlia depth is less than the storage radius.   Chunks get pulled, cached, and gc'd.   And if you're doing enough of them, the gc happens before the cursors get recycled so it starts all over again from the very same peers.  And this keeps happening until the kademlia depth gets back to where it needs to be.
Add some trace statements to the puller if you don't believe this is what is happening.  A while back, I actually had a hacked puller that a) didn't send chunks that weren't within the bin relative to the pull requester, not myself, and b) didn't store locally pulled chunks that were outside of my radius.  This kept the disk churn to a minimum on my end.   But then I realized that (a) was actually benefiting the other nodes which would be good if it is generally deployed, but certainly didn't help on my end.
And the puller really doesn't want to pull high numbered bins from nodes if those chunks are guaranteed to be outside the pulling node's storage radius.  But in fact, it needs to pull from peer bins that are LESS than the pullers storage radius, because that's where the reserve-interested chunks reside over in the peer.
Bottom line:  My bin 8 isn't the same as my peer's bin 8 unless we are in the same depth 8 neighborhood (matching the first 8 bits of the overlay).

We should only be syncing with peers with PO >= storage radius. Any lower, the chunks from these peers do not land in our reserve.

**Comment by ldeffenb** [comment]
Actually, you could just put that comment on #3692

**Comment by zelig** [comment]
It eludes me how such mistakes get into the code. Since when has this bee the casse. 
To make sure it is correct, lets reiterate the desired pull sync strategy

for any node with overlay n and peer with overlay p, let po = PO(n.p),  then
- n must  sync p's own bin po from p  if the peer is outside  your neighbourhood po < storageDepth(n)
- n must sync all of p's PO bins m >= po when the peer is a neighbour po >= storageDepth(n)

**Comment by ldeffenb** [comment]
IMHO, mistakes like this got into the code because the puller and pullsync was written when there was no storage radius, but only a kademlia depth.   When the storage radius finally arrived, the updates were incompletely applied to the puller.   I didn't actually look at the earlier versions of the puller, but I know the puller has given my large pinning nodes fits for a long time before the storage radius arrived in the code.

**Comment by istae** [comment]
> * n must sync p's own bin po from p  if the peer is outside  your neighbourhood po < storageDepth(n)
@zelig what do you mean by this?

**Comment by zelig** [comment]
> > * n must sync p's own bin po from p  if the peer is outside  your neighbourhood po < storageDepth(n)
> 
> @zelig what do you mean by this?

if your peer is not in your neighbourhood then you only pull one bin, the one that contains chunks closer to you than them, i.e. PO bin  po where po =  PO(n,p)
